### DISCUSSION TITLE
Hermione expect

### DISCUSSION THREAD

**Discussion by MetallIdea** [gemini-testing/testplane#887] [Q&A] [answered]
С версии 5.6.0 появился глобальный expect, который конфликтует с expect из jest.
В частности нельзя переопределить expect с помощью @testing-library/jest-dom
Можно ли вынести expect в отдельную либу например hermione/expect только для тех, кому она нужна?

**Comment by KuznetsovRoman** [comment]
Привет. Вынести `expect` в отдельную либу в текущих версиях мы точно не можем, потому что это мажор.

Можете поподробнее рассказать про свой кейс? Этот `expect` вам нужен в Гермиона-тестах? Пробовали ли вы использовать вот такое решение? https://www.npmjs.com/package/@testing-library/jest-dom#with-another-jest-compatible-expect

**Comment by MetallIdea** [comment]
Там не проблема в расширении методов - jest тесты корректно отрабатывают. Проблема именно в типизации, что типы из testing-library не подтягиваются после обновления Гермионы. Попробую этот способ.

**Reply by KuznetsovRoman** [reply]
Покопал `expect-webdriverio`, в доке написано, что можно добавлять и свои тайпинги, это делается как с дефолтным `jest-expect`: https://webdriver.io/docs/api/expect-webdriverio/#adding-your-own-matchers

В `@testing-library/jest-dom` описано, как можно вытащить матчеры для слияния с другим jest-совместимым `expect`: https://github.com/testing-library/jest-dom?tab=readme-ov-file#with-another-jest-compatible-expect

Аналогичным образом вытаскиваю тайпинги. Получается что-то такое:
```ts
import type * as matchers from "@testing-library/jest-dom/types/matchers";

declare global {
    namespace ExpectWebdriverIO {
        interface Matchers<R, T> extends matchers.TestingLibraryMatchers<R, T> {}
    }
}
```
В результате тайпинги из `@testing-library/jest-dom` подъезжают в `expect`:
![image](https://github.com/gemini-testing/hermione/assets/54893992/9c3677a4-f77b-4851-bac4-502a60587d8c)

Накидал пример в code sandbox: https://codesandbox.io/p/devbox/compassionate-taussig-fyd3xs?file=%2Fhermione-tests%2Fexample.hermione.ts%3A9%2C1

Тут нужно обратить внимание на папку `hermione-tests` и гермиона-конфиг (который читает тесты из `hermione-tests/**/*.hermione.(t|j)s`

Попробуйте пожалуйста решить проблему с тайпингами этим способом.

**Reply by MetallIdea** [reply]
Это решило ошибку с типами testing-library, но не решило ошибки с методами из jest, например toHaveLength, toBeCalledWith

**Reply by MetallIdea** [reply, marked as answer]
```typescript
// expect.d.ts

import {type TestingLibraryMatchers} from '@testing-library/jest-dom/types/matchers';

declare global {
    namespace ExpectWebdriverIO {
        interface Matchers<R = void, T = {}>
            extends TestingLibraryMatchers<R, T>,
                jest.JestMatchers<T> {}
    }
}
```

а вот так вроде все нормально, спасибо за помощь!
